snapcraft/plugins/nodejs.py

```diff
@@ -135,144 +110,159 @@ def _nodejs_tar(self):
+            self.run(cmd + ["pack", "--filename", package_tar_path], rootdir)
         else:
-            yarn_add = yarn_cmd + ["add"]
-        for pkg in self.options.node_packages:
-            self.run(yarn_add + [pkg] + flags, cwd=rootdir)
-
-        # local packages need to be added as if they were remote, we
-        # remove the local package.json so `yarn add` doesn't pollute it.
-        if os.path.exists(self._source_package_json):
-            with contextlib.suppress(FileNotFoundError):
-                os.unlink(os.path.join(rootdir, "package.json"))
+            self.run(cmd + ["pack"], rootdir)
+
+        package_dir = os.path.join(rootdir, "package")
+        package_tar = sources.Tar(package_tar_path, rootdir)
+        package_tar.file = package_tar_path
+        os.makedirs(package_dir, exist_ok=True)
+        package_tar.provision(package_dir)
+
+        with contextlib.suppress(FileNotFoundError):
             shutil.copy(
-                self._source_package_json, os.path.join(rootdir, "package.json")
-            )
-            self.run(yarn_add + ["file:{}".format(rootdir)] + flags, cwd=rootdir)
-
-        # npm run would require to bring back package.json
-        if self.options.npm_run and os.path.exists(self._source_package_json):
-            # The current package.json is the yarn prefilled one.
-            with contextlib.suppress(FileNotFoundError):
-                os.unlink(os.path.join(rootdir, "package.json"))
-            os.link(self._source_package_json, os.path.join(rootdir, "package.json"))
-        for target in self.options.npm_run:
-            self.run(
-                yarn_cmd + ["run", target],
-                cwd=rootdir,
-                env=self._build_environment(rootdir),
+                os.path.join(rootdir, "yarn.lock"),
+                os.path.join(package_dir, "yarn.lock"),
             )
-        return self._get_installed_node_packages("npm", self.installdir)
 
-    def _get_installed_node_packages(self, package_manager, cwd):
+        self.run(cmd + ["install"] + flags, package_dir)
+
+        return package_dir
+
+    def run(self, cmd, rootdir):
+        super().run(cmd, cwd=rootdir, env=self._build_environment())
+
+    def run_output(self, cmd, rootdir):
+        return super().run_output(cmd, cwd=rootdir, env=self._build_environment())
+
+    def _build_environment(self):
+        env = os.environ.copy()
+        npm_bin = os.path.join(self._npm_dir, "bin")
+
+        if env.get("PATH"):
+            new_path = "{}:{}".format(npm_bin, env.get("PATH"))
+        else:
+            new_path = npm_bin
+
+        env["PATH"] = new_path
+        return env
+
+    def _get_package_json(self, rootdir):
+        with open(os.path.join(rootdir, "package.json")) as json_file:
+            return json.load(json_file)
+
+    def _get_installed_node_packages(self, cwd):
+        # There is no yarn ls
+        cmd = [os.path.join(self._npm_dir, "bin", "npm"), "ls", "--json"]
```

Comment:
This seems to fail on my machine that doesn't have `npm` installed globally, fresh OS install. Fixed by installing `npm` globally.

Reply:
Looks like peer dependencies complaining made this command fail